unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,699 @@
+use cid::Cid;
+
+use crate::pb::{FlatUnixFs, PBLink, UnixFs, UnixFsType};
+use quick_protobuf::{MessageWrite, Writer};
+use std::borrow::Cow;
+use std::fmt;
+
+use sha2::{Digest, Sha256};
+
+/// File tree builder. Implements `Default` which tracks the recent defaults.
+///
+/// Custom file tree builder can be created with [`FileAdder::builder()`] and configuring the
+/// chunker and collector.
+///
+/// Current implementation maintains an internal buffer for the block creation and uses a
+/// non-customizable hash function to produce Cid version 0 links. Currently does not support
+/// inline links.
+#[derive(Default)]
+pub struct FileAdder {
+    chunker: Chunker,
+    collector: Collector,
+    block_buffer: Vec<u8>,
+    // all unflushed links as a flat vec; this is compacted as we grow and need to create a link
+    // block for the last N blocks, as decided by the collector
+    unflushed_links: Vec<(usize, Cid, u64, u64)>,
+}
+
+impl fmt::Debug for FileAdder {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "FileAdder {{ chunker: {:?}, block_buffer: {}/{}, unflushed_links: {} }}",
+            self.chunker,
+            self.block_buffer.len(),
+            self.block_buffer.capacity(),
+            LinkFormatter(&self.unflushed_links),
+        )
+    }
+}
+
+struct LinkFormatter<'a>(&'a [(usize, Cid, u64, u64)]);
+
+impl fmt::Display for LinkFormatter<'_> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut iter = self.0.iter().peekable();
+
+        write!(fmt, "[")?;
+
+        let mut current = match iter.peek() {
+            Some((depth, ..)) => depth,
+            None => return write!(fmt, "]"),
+        };
+
+        let mut count = 0;
+
+        for (next_depth, ..) in iter {
+            if current == next_depth {
+                count += 1;
+            } else {
+                write!(fmt, "{}/", count)?;
+                for _ in *next_depth..(*current - 1) {
+                    write!(fmt, "0/")?;
+                }
+                count = 1;
+                current = next_depth;
+            }
+        }
+
+        write!(fmt, "{}]", count)
+    }
+}
+
+/// Convinience type to facilitate configuring FileAdders
+#[derive(Default)]
+pub struct FileAdderBuilder {
+    chunker: Chunker,
+    collector: Collector,
+}
+
+impl FileAdderBuilder {
+    /// Configures the builder to use the given chunker.
+    pub fn set_chunker(&mut self, chunker: Chunker) -> &mut Self {
+        self.chunker = chunker;
+        self
+    }
+
+    /// Configures the builder to use the given collector or layout.
+    pub fn set_collector(&mut self, collector: Collector) -> &mut Self {
+        self.collector = collector;
+        self
+    }
+
+    /// Returns a new FileAdder
+    pub fn build(&self) -> FileAdder {
```

Comment:
I'm used to `build` consuming `self`, but if we want to reuse them then it's fine

Reply:
Forgot this much from your previous comments. I guess this comes down to either having:

1. the `with_` methods that are `with_foo(self, foo: Foo) -> Self` AND moving `build(self) -> Target`
2. the `set_` methods that are `set_foo(&mut self, foo: Foo) -> &mut Self` AND borrowing+cloning `build(&self) -> Target`

So ... The first one has previously been used in this crate but you commented then that `&mut self` is preferred. I always lean towards the second option. Am I missing some better third way?

Reply:
Isn't this the final function in the builder pattern, like a `finish`? I think that other case was a bit different, but no longer remember how exactly :stuck_out_tongue:.

Reply:
I think the previous one was related to [`IdleFileVisit::with_target_range`](https://docs.rs/ipfs-unixfs/0.0.1/ipfs_unixfs/file/visit/struct.IdleFileVisit.html#method.with_target_range) discussed in https://github.com/rs-ipfs/rust-ipfs/pull/176/files#r436010123

Reply:
ah, but that one was `with_x`, which usually doesn't indicate a move, but rather mutation; `build`/`finish` totally suggest finality (e.g. consuming the argument) IMO

Reply:
I am leaning towards making this the second variant, if for nothing else than to keep it inline with the little we already have.

Reply:
Done